plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php

```diff
@@ -0,0 +1,317 @@
+	 * - 'info_link': if present, a "More information" link will be included in the notice.
+	 *
+	 * Additionally, $metadata can contain a 'plugin_name' key to be used to check if the plugin installed
+	 * successfully; if not present, the plugin name will be inferred from the diff of the installed plugins list
+	 * before and after the installation process.
+	 *
+	 * Information about plugins successfully installed with this method will be kept in an option named
+	 * 'woocommerce_autoinstalled_plugins'. Keys will be the plugin name and values will be associative arrays
+	 * with these keys: 'plugin_name', 'version', 'date' and 'metadata' (same meaning as in the returned array).
+	 *
+	 * A log entry will be created with the result of the process and all the installer messages
+	 * (source: 'plugin_auto_installs'). In multisite this log entry will be created on each site.
+	 *
+	 * The returned array will contain the following (only 'install_ok' and 'messages' if the installation fails):
+	 *
+	 * - 'install_ok', a boolean.
+	 * - 'messages', all the messages generated by the installer.
+	 * - 'plugin_name', useful if it wasn't supplied in $metadata.
+	 * - 'version', of the plugin that has been installed.
+	 * - 'date', ISO-formatted installation date.
+	 * - 'metadata', as supplied (except the 'plugin_name' key) and only if not empty.
+	 *
+	 * @param string $plugin_url URL or file path of the plugin to install.
+	 * @param array  $metadata Metadata to store if the installation succeeds.
+	 * @return array Information about the installation result.
+	 */
+	public function install_plugin( string $plugin_url, array $metadata = array() ): array {
+		$this->installing_plugin = true;
+		try {
+			return $this->install_plugin_core( $plugin_url, $metadata );
+		} finally {
+			$this->installing_plugin = false;
+		}
+	}
+
+	/**
+	 * Core version of 'install_plugin' (it doesn't handle the $installing_plugin flag).
+	 *
+	 * @param string $plugin_url URL or file path of the plugin to install.
+	 * @param array  $metadata Metadata to store if the installation succeeds.
+	 * @return array Information about the installation result.
+	 */
+	private function install_plugin_core( string $plugin_url, array $metadata ): array {
+		wp_clean_plugins_cache();
+
+		$plugin_name = $metadata['plugin_name'] ?? null;
+		unset( $metadata['plugin_name'] );
+		if ( is_null( $plugin_name ) ) {
+			$old_plugin_names = array_keys( get_plugins() );
+		} elseif ( ! is_null( $this->get_installed_plugin_info( $plugin_name ) ) ) {
+			return array(
+				'install_ok'         => false,
+				'installer_messages' => __( 'Plugin is already installed', 'woocommerce' ),
+			);
+		}
+
+		include_once ABSPATH . 'wp-admin/includes/file.php';
+		include_once ABSPATH . 'wp-admin/includes/misc.php';
+		include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+		include_once ABSPATH . 'wp-admin/includes/class-automatic-upgrader-skin.php';
```

Comment:
For extra safety I believe we could _detect_ whether the classes we need from these files already exist before attempting the include, for which we should also check that the file actually exists, [similar to what WP-CLI does](https://github.com/wp-cli/wp-cli/blob/4c68c264b3aa5a987291f36882ad71387c28f997/php/utils-wp.php#L156-L166).

Even though we have an `include_once()` it _might_ be the case that those classes change locations in WP at some point in the future. Maybe I'm worrying too much but it wouldn't hurt to cleanly bail out if we can't properly load the files and classes we need for this function.

Reply:
Ok, I'm now checking if the classes exist before executing the includes.